### PR TITLE
ROX-13838: deprovision disgraced centrals

### PR DESCRIPTION
**Pull request by 0x656b694d**
Adjust query to mark disgraced centrals to deprovision.

### CODE REVIEW

**Comment by 0x656b694d** [comment]
Current dependencies on/for this PR:

  * main
    * **PR #1248**
      * **PR #1250** 👈

File: internal/dinosaur/pkg/services/dinosaur.go [hunk 1]
```diff
@@ -53,6 +50,8 @@ const DinosaurRoutesActionCreate DinosaurRoutesAction = "CREATE"
 // DinosaurRoutesActionDelete ...
 const DinosaurRoutesActionDelete DinosaurRoutesAction = "DELETE"
 
+const gracePeriod = 12 * 24 * time.Hour
```
Comment (ivan-degtiarenko) [inline internal/dinosaur/pkg/services/dinosaur.go:53]: 12 days?

Comment (0x656b694d) [reply internal/dinosaur/pkg/services/dinosaur.go:53]: Random enough! A typo, I meant 14.

File: internal/dinosaur/pkg/services/dinosaur.go [hunk 2]
```diff
@@ -519,9 +518,12 @@ func (k *dinosaurService) DeprovisionDinosaurForUsers(users []string) *errors.Se
 func (k *dinosaurService) DeprovisionExpiredDinosaurs(dinosaurAgeInHours int) *errors.ServiceError {
 	now := time.Now()
 	dbConn := k.connectionFactory.New().
-		Model(&dbapi.CentralRequest{}).
+		Model(&dbapi.CentralRequest{})
+
+	dbConn = dbConn.Where(dbConn.
 		Where("instance_type = ?", types.EVAL.String()).
 		Where("created_at  <=  ?", now.Add(-1*time.Duration(dinosaurAgeInHours)*time.Hour)).
```
Comment (ivan-degtiarenko) [inline internal/dinosaur/pkg/services/dinosaur.go:525]: I don't think we should reuse this function. Then if we enable `EnableDeletionOfExpiredCentral` our instances will have a fixed lifetime as a side effect.

Let's introduce separate function

Comment (0x656b694d) [reply internal/dinosaur/pkg/services/dinosaur.go:525]: Isn't it what we actually want? Deleting expired normal and eval instances?

Comment (ivan-degtiarenko) [reply internal/dinosaur/pkg/services/dinosaur.go:525]: Did we decide that `eval` instances(or normal instances) would have a fixed lifetime? If so, can you point me to where this decision was made?

Comment (0x656b694d) [reply internal/dinosaur/pkg/services/dinosaur.go]: Refactored.

File: internal/dinosaur/pkg/services/dinosaur.go [hunk 3]
```diff
@@ -516,13 +515,18 @@ func (k *dinosaurService) DeprovisionDinosaurForUsers(users []string) *errors.Se
...
 	now := time.Now()
-	dbConn := k.connectionFactory.New().
-		Model(&dbapi.CentralRequest{}).
-		Where("instance_type = ?", types.EVAL.String()).
-		Where("created_at  <=  ?", now.Add(-1*time.Duration(dinosaurAgeInHours)*time.Hour)).
-		Where("status NOT IN (?)", dinosaurDeletionStatuses)
+	dbConn := k.connectionFactory.New().Model(&dbapi.CentralRequest{}).
+		Where("expired_at IS NOT NULL").Where("expired_at < ?", now.Add(-gracePeriod))
+
+	if lifespan := k.dinosaurConfig.CentralLifespan; lifespan.EnableDeletionOfExpiredCentral {
```
Comment (ivan-degtiarenko) [inline internal/dinosaur/pkg/services/dinosaur.go:523]: nit:
```suggestion
	if k.dinosaurConfig.CentralLifespan.EnableDeletionOfExpiredCentral {
```

**Review by ivan-degtiarenko — APPROVED** [review]
LGTM considering manual test will run successfully

**Review by ivan-degtiarenko — APPROVED** [review]
LGTM considering manual test will run successfully

**Pull request merged**